[Nicosia] Add support for translate/rotate/scale animations

obyknovenius opened the pull request:
#### 9f21b9e439f93986a5e4f7dc52b14b156d0f6677
<pre>
[Nicosia] Add support for translate/rotate/scale animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=269288">https://bugs.webkit.org/show_bug.cgi?id=269288</a>

Reviewed by Nikolas Zimmermann.

Nicosia already supports accelerated transform animations.
This patch implements translate, rotate, and scale animations.

It also ensures that when multiple animations are applied,
they run in the correct order required the spec [1]:
- translate
- rotate
- scale
- transform

[1] <a href="https://drafts.csswg.org/css-transforms-2/#ctm">https://drafts.csswg.org/css-transforms-2/#ctm</a>

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
(Nicosia::Animation::apply):
(Nicosia::Animation::applyInternal):
(Nicosia::Animations::apply):
(Nicosia::Animations::hasRunningTransformAnimations const):
(Nicosia::Animation::applyKeepingInternalState): Deleted.
(Nicosia::Animations::applyKeepingInternalState): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.h:
(Nicosia::Animations::setTranslate):
(Nicosia::Animations::setRotate):
(Nicosia::Animations::setScale):
(Nicosia::Animations::setTransform):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::syncAnimations):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::didChangeAnimations):
(WebCore::CoordinatedGraphicsLayer::addAnimation):
(WebCore::CoordinatedGraphicsLayer::transformRelatedPropertyDidChange):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:

Canonical link: <a href="https://commits.webkit.org/275168@main">https://commits.webkit.org/275168@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e86904efd18fa7b90f9c2a15c0ef01170242cdb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40895 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19908 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43273 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43453 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36986 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43202 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17239 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33893 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41469 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16852 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35250 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14516 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14621 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36240 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44748 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37106 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36551 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40301 "Found 1 new test failure: webanimations/accelerated-animation-with-easing.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15710 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12900 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38669 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17329 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9221 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17380 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16973 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
```diff
@@ -367,8 +372,45 @@ void Animations::resume()
 void Animations::apply(Animation::ApplicationResult& applicationResults, MonotonicTime time)
 {
-    for (auto& animation : m_animations)
-        animation.apply(applicationResults, time);
```
You are now iterating the animations once, and divide them into 5 buckets, storing adresses of the objects seperated. Then loop over each of those containers. Since typically you have a few animations only registered, I think you are better off if you just loop once per animation type.

Also will be easier to read.

```
auto applyAnimationType = [&](auto type) {
    for (auto& animatin : m_animations) {
        if (animation.keyframes().property() == type)
            animation.apply(applicationResults, time)
    }
};

applyAnimationType(AnimatedProperty::Translate);
applyAnimationType(AnimatedProperty::Scale);
...
```

obyknovenius replied:
I was thinking about this approach. But it would mean to iterate all animations five times. Whereas in the current solution, we visit every animation only twice. But I'm okay with either implementation.

obyknovenius replied:
Also, this exact implementation won't work because as the last step, we want to run all animation types except transformation-like animations without a specific order:
- `AnimatedProperty::Opacity`
- `AnimatedProperty::BackgroundColor`
- `AnimatedProperty::Filter`
- `AnimatedProperty::WebkitBackdropFilter`

nikolaszimmermann replied:
That's correct the lambda needs to be generalized for that case. Anyhow, I have no strong opinion to the implementation, at the moment you are allocating five temporary vectors - probably all on stack so either way is ok, unless this shows up in profiling 😏

nikolaszimmermann reviewed (changes requested):
I think it should say what is fixed by this :-)

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
```diff
@@ -306,9 +306,15 @@ Seconds Animation::computeTotalRunningTime(MonotonicTime time)
+    case AnimatedProperty::Transform: {
+        ASSERT(applicationResults.transform);
+        auto transform = applyTransformAnimation(static_cast<const TransformAnimationValue&>(from).value(), static_cast<const TransformAnimationValue&>(to).value(), progress, m_boxSize);
+        applicationResults.transform = applicationResults.transform->multiply(transform);
```
The assignment is unncessary, multiple returns '*this' already, so the multiplication is performed in-place.

obyknovenius replied:
Right.

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
```diff
@@ -365,16 +371,62 @@ void Animations::resume()
+
+        if (translateAnimations.isEmpty())
+            applicationResults.transform->multiply(m_translate);
+        else
```
Missing braces here and in the other multi-line else branches below.

obyknovenius replied:
Added braces.

nikolaszimmermann commented on Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
```diff
@@ -365,16 +371,62 @@ void Animations::resume()
+    }
+
+    if (!translateAnimations.isEmpty() || !rotateAnimations.isEmpty() || !scaleAnimations.isEmpty() || !transformAnimations.isEmpty()) {
+        applicationResults.transform = TransformationMatrix();
```
Now you know applicationResults.transform is no longer std::nullopt, how about adding:
```
auto& transform = applicationResults.transform.value();
```
and avoid the `applicationResults.transform->multiply(...)` calls below and simly use `transform.multiply(...)` - wdyt?

obyknovenius replied:
I'm not sure about this one. I prefer to be more explicit that we deal with `applicationResults` here.

nikolaszimmermann reviewed (approved):
Great patch, r=me.

obyknovenius committed:
[Nicosia] Add support for translate/rotate/scale animations
https://bugs.webkit.org/show_bug.cgi?id=269288

Reviewed by Nikolas Zimmermann.

Nicosia already supports accelerated transform animations.
This patch implements translate, rotate, and scale animations.

It also ensures that when multiple animations are applied,
they run in the correct order required the spec [1]:
- translate
- rotate
- scale
- transform

[1] https://drafts.csswg.org/css-transforms-2/#ctm

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.cpp:
(Nicosia::Animation::apply):
(Nicosia::Animation::applyInternal):
(Nicosia::Animations::apply):
(Nicosia::Animations::hasRunningTransformAnimations const):
(Nicosia::Animation::applyKeepingInternalState): Deleted.
(Nicosia::Animations::applyKeepingInternalState): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaAnimation.h:
(Nicosia::Animations::setTranslate):
(Nicosia::Animations::setRotate):
(Nicosia::Animations::setScale):
(Nicosia::Animations::setTransform):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::syncAnimations):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::didChangeAnimations):
(WebCore::CoordinatedGraphicsLayer::addAnimation):
(WebCore::CoordinatedGraphicsLayer::transformRelatedPropertyDidChange):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:

Canonical link: https://commits.webkit.org/275168@main
